modules/migrations/base/downloader.go

```diff
@@ -174,3 +190,86 @@ func (d *RetryDownloader) GetPullRequests(page, perPage int) ([]*PullRequest, er
 	}
 	return nil, err
 }
+
+var (
+	_ Downloader = &LimitDownloader{}
+)
+
+// LimitDownloader limit the downloads
+type LimitDownloader struct {
+	Downloader
+	startTime time.Time // start time
+}
+
+// NewLimitDownloader creates a limit downloader
+func NewLimitDownloader(downloader Downloader) *LimitDownloader {
+	return &LimitDownloader{
+		Downloader: downloader,
+		startTime:  time.Now(),
+	}
+}
+
+// GetRequestTimes returns request times
+func (d *LimitDownloader) GetRequestTimes() int {
+	return d.Downloader.GetRequestTimes()
+}
+
+// GetRequestLimit returns the limitation of the http request times per seconds.
+func (d *LimitDownloader) GetRequestLimit() float32 {
+	return d.Downloader.GetRequestLimit()
+}
+
+func (d *LimitDownloader) sleep() {
+	secs := int(time.Since(d.startTime) / time.Second)
+	if secs > 0 && float32(d.GetRequestTimes()/secs) > d.GetRequestLimit() {
+		time.Sleep(time.Second * time.Duration(int(float32(d.GetRequestTimes())/d.GetRequestLimit())-secs))
```

Comment:
I see some problems with this approach:

* This assumes the rate "counter" in GitHub is uniform, while in fact it could be that it's reset at the turn of the hour.
* This will artificially slow down any migration with less than 5,000 requests (e.g. perhaps a 4,500 request migration could be resolved within minutes, but now it will take nearly one hour to resolve).
* This method can't possibly take any previous attempts into account, so it may fail repeatedly despite all efforts.

IMHO I think it's better to take [the actual response from GitHub](https://developer.github.com/v3/#rate-limiting) for these calculations instead of having our own:

![image](https://user-images.githubusercontent.com/18600385/70772349-fd27f700-1d52-11ea-9ca5-3370c8fb1762.png)

There's also the [rate limit API](https://developer.github.com/v3/rate_limit/) we should call before starting making queries.